src/plugins/media/media.js

```diff
@@ -0,0 +1,261 @@
+ *    for individual steps and media.
+ *
+ *    Examples:
+ *    - data-media-autostart="true" data-media-autostop="true": start media on enter, stop on leave, 
+ *      restart from beginning when re-entering the step.
+ *
+ *    - data-media-autostart="true" data-media-autopause="true": start media on enter, pause on leave, 
+ *      resume on re-enter
+ *
+ *    - data-media-autostart="true" data-media-autostop="true" data-media-autopause="true": start 
+ *      media on enter, stop on leave (stop overwrites pause).
+ *
+ *    - <div id="impress" data-media-autostart="true"> ... <div class="step" data-media-autostart="false">
+ *      All media is startet automatically on all steps except the one that has the data-media-autostart="false"
+ *      attribute.
+ *
+ *  - Pro tip: Use <audio onended="impress().next()"> or <video onended="impress().next()"> to proceed to the 
+ *    next step automatically, when the end of the media is reached.
+ *    
+ *
+ * Copyright 2018 Holger Teichert (@complanar)
+ * Released under the MIT license.
+ */
+/* global window, document */
+
+(function (document, window) {
+    "use strict";
+    var root, api, gc, gcAttributes;
+    
+    gcAttributes = [];
+    
+    // function names
+    var enhanceMediaNodes,
+        enhanceMedia,
+        removeMediaClasses,
+        onStepenter,
+        onStepleave,
+        onPlay,
+        onPause,
+        onEnded,
+        garbage;
+    
+    document.addEventListener("impress:init", function (event) {
+        root = event.target;
+        api = event.detail.api;
+        gc = api.lib.gc;
+        
+        if (root.dataset.mediaAutoplay === "" || root.dataset.mediaAutoplay === "true") {
+            root.mediaAutoplay = true;
+        } else {
+            root.mediaAutoplay = false;
+        }
+        if (root.dataset.mediaAutostop === "" || root.dataset.mediaAutostop === "true") {
+            root.mediaAutostop = true;
+        } else {
+            root.mediaAutostop = false;
+        }
+        if (root.dataset.mediaAutopause === "" || root.dataset.mediaAutopause === "true") {
+            root.mediaAutopause = true;
+        } else {
+            root.mediaAutopause = false;
+        }
+        
+        // this *must* be called only after setting the autopause, autoplay and autostop values
+        enhanceMedia();
+        
+        gc.pushCallback(garbage);
+    }, false);
+    
+    garbage = function () {
+        var elements, el, i;
+        removeMediaClasses();
+        delete root.mediaAutoplay;
+        delete root.mediaAutopause;
+        delete root.mediaAutopstop;
+        elements = document.querySelectorAll("#root, .step, audio, video");
+        for (i = 0; i < elements.length; i += 1) {
+            el = elements[i];
+            delete el.mediaAutoplay;
+            delete el.mediaAutopause;
+            delete el.mediaAutostop;
+        }
```

Comment:
The above attributes are created by the user. You should not delete them.

The below are automatically created by yourself, those you should delete.

In other words, you can delete the above code block.

Reply:
Your comment confuses me. Are custom properties I add to a DOM object automatically transferred into HTML attributes?

`el.dataset.mediaAutoplay` refers to the user defined HTML attributes
`el.mediaAutoplay` refers to my defined properties.

This should be comparable to the way the rel plugin works, isn’t it? Or am I missing anything important here?

Reply:
Noting to myself I have replied to this in another thread.